broadcast/broadcast.go

```diff
@@ -49,17 +56,49 @@ func NewBroadcaster(n int) *Broadcaster {
 	return &Broadcaster{capacity: n}
 }
 
-// Send broadcasts a message to the channel.
+// SendWithTimeout broadcasts a message to each listener's channel.
 // Sending on a closed channel causes a runtime panic.
-func (b *Broadcaster) Send(v interface{}) {
+// This method blocks for a duration of up to `timeout` on each channel.
+// Returns error(s) if it is unable to send on a given listener's channel within `timeout` duration.
+func (b *Broadcaster) SendWithTimeout(v interface{}, timeout time.Duration) error {
 	b.m.Lock()
 	defer b.m.Unlock()
 	if b.closed {
 		panic("broadcast: send after close")
 	}
-	for _, l := range b.listeners {
-		l <- v
+	var result *multierror.Error
+	for id, l := range b.listeners {
+		select {
+		case l <- v:
+			// Success!
+		case <-time.After(timeout):
+			err := fmt.Sprintf("unable to send to listener '%d'", id)
+			result = multierror.Append(result, errors.New(err))
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+// Send broadcasts a message to each listener's channel.
+// Sending on a closed channel causes a runtime panic.
+// This method is non-blocking, and will return errors if unable to send on a given listener's channel.
+func (b *Broadcaster) Send(v interface{}) error {
+	b.m.Lock()
+	defer b.m.Unlock()
+	if b.closed {
+		panic("broadcast: send after close")
```

Comment:
Deciding to panic is aligned with what Go does when sending it to a closed channel, so it makes sense.
I'm wondering if in this kind of `Broadcaster` abstraction would also see this situation as an irrecoverable error. Not sure though.

Reply:
Yeh, very good question. My preference would actually be to _not_ panic here, but this is what came with https://godoc.org/github.com/tjgq/broadcast (which this is based on).

Reply:
Then again, I'm not a fan of silently not doing what the called asked 😬, so if we don't panic here, should we return an error instead? Seems like a reasonable thing to do?

Reply:
Yes, I was thinking of returning a defined error. 
Not doing anything silently is bad.

Reply:
See f3146e139162d21713b4c993e6ebed7aaad77207

Reply:
LGTM!